### PR TITLE
Feat/#5/map collections

### LINKED ISSUE #5: Implement Map collections

### PR DESCRIPTION
**Pull request by maciej-sz**
## Summary by CodeRabbit

- **Documentation**
  - Enhanced the README with clearer guidance on the Comfy Gopher toolset and detailed objectives for the Collections package.

- **New Features**
  - Introduced a utility to count occurrences within collections.
  - Added enhanced collection operations that now return additional result details (e.g., removed elements).
  - Implemented a new map structure with improved key-value pair handling and additional methods for manipulation.
  - Added new methods for sequences and maps to facilitate better value management and retrieval.

- **Refactor**
  - Standardized type definitions and method signatures across collection-related APIs for improved consistency and error handling.
  - Updated internal structures to simplify and enhance functionality in various collection types.

- **Tests**
  - Expanded and refined test scenarios to ensure robust handling of duplicate items and various edge cases.
  - Introduced new tests for the updated map and sequence functionalities, including value counting and error handling.
  - Added tests for the new `valuesCounter` functionality to validate counting operations.

- **Chores**
  - Updated build configurations to streamline resource management.

### CODE REVIEW

**Comment by maciej-sz** [comment]
@coderabbitai review

File: functions_internal.go [hunk 1]
```diff
@@ -143,95 +168,113 @@
...
 	first := true
-	coll.EachRev(func(i int, v V) {
+	for i := len(s) - 1; i >= 0; i-- {
 		if first {
-			acc = v
+			acc = s[i]
 			first = false
 		} else {
-			acc = reducer(acc, i, v)
+			acc = reducer(acc, i, s[i])
```
[the remark this answers is not in the document]
Comment (maciej-sz) [reply functions_internal.go:220]: This will be handled when resolving other issues..

File: definitions.go [hunk 2]
```diff
@@ -8,32 +8,37 @@ import (
...
-	// ErrKeyAlreadyExists
+	// ErrOutOfBounds is returned when an operation is performed on an index that is out of bounds.
 	ErrOutOfBounds = errors.New("index out of bounds")
-	// ErrValueNotFound
+	// ErrEmptyCollection is returned when an operation is performed on an empty collection.
 	ErrEmptyCollection = fmt.Errorf("%w: collection is empty", ErrOutOfBounds)
-	// ErrKeyAlreadyExists
+	// ErrValueNotFound is returned when a value is not found in a collection.
 	ErrValueNotFound = errors.New("value not found")
 )
```
[the remark this answers is not in the document]
Comment (maciej-sz) [reply definitions.go:17]: It is as intended: `ErrEmptyCollection` is a special case of `ErrOutOfBounds`.
